### DISCUSSION TITLE
Re-using components as field

### DISCUSSION THREAD

**Discussion by jaepetto** [react-hook-form/react-hook-form#3901] [Q&A] [answered]
Hi,

(this is a newbie question so please be gentle :-) )

I have a bunch of forms that are using the same fields (same logic, validation, ...). Typically, the code of one of the forms looks like below:
```javascript
import { useForm } from "react-hook-form";
import VisibilitySelector from '../Common/VisibilitySelector';
import React from 'react'

export default function EditEducation({
  education,
  educationsUpdateFn,
  cancelEditionFn,
}) {
  const { register, handleSubmit, reset, errors } = useForm();

  const onSubmit = (data) => {
    educationsUpdateFn(data);
    reset();
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <input
        type="hidden"
        name="id"
        ref={register({ required: true })}
        defaultValue={education.id}
      />
      <VisibilitySelector />
      <input type="submit" value="Save" />
      <button onClick={cancelEditionFn}>Cancel</button>
    </form>
  );
}
```

And the visibility selector looks like:
```javascript
import React from "react";
import { useForm } from "react-hook-form";

function VisibilitySelector() {
  const { register, errors, watch } = useForm();

  console.log(watch("visibility"))
  return (
    <>
      <label htmlFor="visibility">Visibility</label>
      <select
        name="visibility"
        id="visibility"
        ref={register({ required: true })}
      >
        <option value="">--- select one ---</option>
        <option value="all">all years</option>
        <option value="2020">2020</option>
        <option value="2021">2021</option>
        <option value="2022">2022</option>
        <option value="2023">2023</option>
        <option value="2024">2024</option>
        <option value="2025">2025</option>
      </select>
      {errors.visibility?.type === "required" && (
        <span>This field is required</span>
      )}
    </>
  );
}

export default VisibilitySelector;
```

The idea behind this is to be able to re-use the visibility selector across the various forms. Ideally, this component should have it's own logic (especially the validation) and should be fully encapsulated.

This being said, the above code does not work. The validation does not work and the data is not passed to the form.

While reviewing a couple of possible answers and I found this one: https://github.com/react-hook-form/react-hook-form/discussions/2711

It points to the documentation about connected forms (https://react-hook-form.com/advanced-usage#ConnectForm) but I cannot figure the way to adapt the example to my situation. Furthermore, I believe the Smart Form Component (https://react-hook-form.com/advanced-usage#SmartFormComponent) would ne more appropriate.

Could you please point me to what is the recommended way to achieve this reusability / encapsulation?

Thanks in advance,

Emmanuel

**Comment by bluebill1049** [comment, marked as answer]
I am using the smart form component and combined it with ConnectedForm at work, so any field input at the root level will get registered automatically, when I need to build complex logic I will use connectForm, it gives me flexibility.

**Reply by jaepetto** [reply]
Thanks for the hint. would you have some really simple example available?

**Reply by bluebill1049** [reply]
Don't have them in hand, but what i wrote on the doc is pretty much what i did.

**Reply by jaepetto** [reply]
I just went through the doc again and finally figured something out: it was a lot simpler than I expected. 

I simply had to remove all references (register, errors) to react-hook-form from the sub-component and pass them from the main form...

```javascript
import React from "react";

function VisibilitySelector({register, errors}) {

  return (
    <>
      <label htmlFor="visibility">Visibility</label>
      <select
        name="visibility"
        id="visibility"
        ref={register({ required: true })}
      >
        <option value="">--- select one ---</option>
        <option value="all">all years</option>
        <option value="2020">2020</option>
        <option value="2021">2021</option>
        <option value="2022">2022</option>
        <option value="2023">2023</option>
        <option value="2024">2024</option>
        <option value="2025">2025</option>
      </select>
      {errors.visibility?.type === "required" && (
        <span>This field is required</span>
      )}
    </>
  );
}

export default VisibilitySelector;
```

form:
```javascript
<form onSubmit={handleSubmit(onSubmit)}>
      <input
        type="hidden"
        name="id"
        ref={register({ required: true })}
        defaultValue={education.id}
      />
      <VisibilitySelector register={register} errors={errors} />
      <input type="submit" value="Save" />
      <button onClick={cancelEditionFn}>Cancel</button>
    </form>
```

Thanks a lot for the library and for your time!

cheers,

Emmanuel
